Feature: Implement comprehensive expense history management with instant date picker

lutfarrahmanrifat opened the pull request:
## Overview
This PR introduces a complete expense history management system with improved navigation patterns, centralized date utilities, and an instant date picker for better user experience.

## Key Features

### Expense History Management
- Replace expense history dialog with dedicated ExpenseHistoryScreen
- Add "Latest Expenses" section in Expense tab showing 5 most recent expenses from selected month
- Implement seamless navigation from Latest Expenses to Full History with month context
- Add month-based filtering and sorting for expense history
- Create consistent expense item layout across all screens with proper spacing and typography

### Instant Date Selection
- Replace traditional DatePickerDialog with custom one-touch selection
- Remove OK/Cancel buttons for streamlined user experience
- Implement immediate date application with auto-dismiss functionality
- Maintain fast native performance while improving usability

### Code Quality Improvements
- Centralize date utilities in DateConstants with month name and formatting functions
- Create shared MonthYearSelector component eliminating code duplication across 3 screens
- Implement Flow-based expense observation for real-time UI updates
- Replace hardcoded values with ValidationConstants for better maintainability
- Fix deprecated Date methods with modern Calendar approach

### Technical Enhancements
- Add automatic expense list updates when expenses are deleted from any screen
- Implement parameterized navigation to pass selected month/year between screens
- Remove duplicate functions and unused lifecycle management code
- Establish single source of truth for date formatting throughout the app
- Add proper scrolling support to ExpenseScreen using LazyColumn

## Files Changed
- ExpenseScreen: Add Latest Expenses section and instant date picker
- ExpenseHistoryScreen: New dedicated screen for full expense history
- DateConstants: Centralized date utilities and month name functions
- MonthYearSelector: Shared component eliminating code duplication
- Navigation: Add parameterized routes for expense history
- README: Updated documentation reflecting new features

## Testing
- Verified instant date selection works across multiple selections
- Confirmed real-time expense updates when deleting from different screens
- Tested navigation flow from Latest Expenses to Full History
- Validated month-based filtering and consistent UI layouts

lutfarrahmanrifat committed:
feat: Implement expense history with consistent navigation pattern

- Replace expense history dialog with dedicated ExpenseHistoryScreen
- Add Latest Expenses section to Expense tab showing 5 most recent expenses
- Implement month-based filtering for expense history
- Create consistent expense item layout across all screens:
  * Category name as primary text
  * Description below category (up to 2 lines in history, 1 line in latest)
  * Amount and date grouped on right side
  * Delete functionality with confirmation dialogs
- Add proper scrolling to ExpenseScreen using LazyColumn
- Remove redundant total expenses card (available in Summary tab)
- Update navigation to include ExpenseHistoryScreen route
- Improve spacing and padding for better visual hierarchy
- Add ExpenseHistoryViewModel with month-based data loading
- Update README with new expense management features

lutfarrahmanrifat committed:
refactor: clean up expense history auto-update implementation

- Remove duplicate MonthYearSelector functions, create shared component
- Implement automatic expense list updates using Flow observation
- Replace hardcoded values with ValidationConstants.LATEST_EXPENSES_COUNT
- Fix deprecated Date methods with modern Calendar approach
- Remove unused functions and lifecycle management code
- Add proper month/year formatting for Latest Expenses title
- Improve code maintainability and consistency

lutfarrahmanrifat committed:
feat: centralize month name utilities and pass selected month to expense history

- Add DateConstants.getMonthName() and getMonthNumber() utilities
- Add DateConstants.getMonthYearString() for consistent date formatting
- Refactor ExpenseScreen to use centralized date utilities
- Remove duplicate getMonthName() function from CategoryExpenseDetailScreen
- Update MonthYearSelector to use centralized month utilities
- Implement parameterized navigation to ExpenseHistoryScreen
- Pass selected month/year from Latest Expenses to Full History
- Eliminate ~30+ lines of duplicate month handling code
- Establish single source of truth for all date formatting

lutfarrahmanrifat committed:
feat: implement instant date picker with one-touch selection

- Replace traditional DatePickerDialog with custom instant selection
- Remove OK/Cancel buttons for streamlined UX
- Add setOnDateChangedListener for immediate date application
- Fix multiple selection issue by creating fresh dialog each time
- Maintain fast native performance while improving usability
- Update README with instant date picker and technical improvements
- Add centralized date utilities documentation
- Document reactive updates and shared components

rifatr commented on app/src/main/java/com/example/budget/data/db/AppDatabase.kt:
Why do you need this change?

rifatr commented on app/src/main/java/com/example/budget/ui/expense/ExpenseScreen.kt:
```diff
@@ -76,128 +73,175 @@ fun ExpenseScreen(
+             else {
+                 item {
+                       Text(
+                           text = "No expenses in ${DateConstants.getMonthYearString(uiState.date)}",
```
Declare a variable for the month name instead of duplicating code.

lutfarrahmanrifat replied:
done.

rifatr reviewed (changes requested):
I've commented some stuffs in the code. Please fix those.

lutfarrahmanrifat committed:
docs: Update README with comprehensive feature documentation

- Add character limits for expense descriptions (150) and category names (24)
- Document smart category selection with lexicographical sorting and last selected pre-selection
- Update expense management section with latest expenses preview and full history screen
- Add centralized validation system and DateConstants documentation
- Include real-time cross-screen synchronization features
- Document instant date picker with no OK button requirement
- Update project structure with new components and proper comment alignment
- Add ValidationConstants and CategoryPreferences to architecture overview
- Update daily usage guide with new expense history workflow

lutfarrahmanrifat committed:
Date constant function updated for data consistency

lutfarrahmanrifat commented on app/src/main/java/com/example/budget/ui/budget/BudgetScreen.kt:
```diff
@@ -2,7 +2,9 @@ package com.example.budget.ui.budget
 import com.example.budget.ui.components.BeautifulSelector
+import com.example.budget.ui.components.MonthYearSelector
 import com.example.budget.ui.components.ConfirmationMessage
+import com.example.budget.data.ValidationConstants
```
For supporting category name length validation

lutfarrahmanrifat commented on app/src/main/java/com/example/budget/data/DateConstants.kt:
```diff
@@ -52,4 +52,38 @@ object DateConstants {
+     * @return Month number (1-12, where 1 = January)
+     */
+    fun getMonthNumber(monthName: String): Int {
+        return MONTHS.find { it.first == monthName }?.second ?: 1
```
Updated.

lutfarrahmanrifat commented on README.md:
Updated

lutfarrahmanrifat committed:
refactor: Update minimum API level and optimize ExpenseScreen

- Update minimum SDK from 24 to 26 (Android 8.0+) for modern API support
- Fix DatePicker setOnDateChangedListener API compatibility issue
- Remove code duplication in ExpenseScreen DateSelector
- Optimize month-year string calculation with single variable reuse
- Update README documentation to reflect new minimum API requirements
- Clean up calendar instance creation with Kotlin apply scope function
- Maintain instant date selection functionality with cleaner implementation
